naptime/src/main/scala/org/coursera/naptime/access/AccessControl.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2016 Coursera Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.coursera.naptime.access
+
+import org.coursera.naptime.NaptimeActionException
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+/**
+ * Control access to a resource method. Arbitrary, asynchronous logic may be executed, but
+ * successful access grant must produce an authentication object of type [[A]].
+ *
+ * @tparam R request data (for example, for Play HTTP requests, this is
```

Comment:
I definitely like the goal of abstracting away from HTTP. I guess I'm not convinced leaving an unbounded, contravariant type parameter is the right way to go. I'd rather defer introducing this trait until after we have a better sense on what interface we're actually trying to provide in lieu of Play-HTTP. (i.e. no current uses of `HeaderAccessControl` can migrate to `AccessControl`.)

Additionally, I'm not sure I like the name of the trait either. :-( There's a lot more to access control than Auth-n/z at the request level. There's field-level access control, and response element-level access control. I don't think these sorts of access control applications would fit under the same interface.